src/main/java/org/rumbledb/runtime/control/TypeswitchRuntimeIterator.java

```diff
@@ -45,16 +45,25 @@ public TypeswitchRuntimeIterator(
     @Override
     public void open(DynamicContext context) {
         super.open(context);
+        if(this.matchingIterator != null)
+        {
+            this.matchingIterator.close();
+        }
+        this.matchingIterator = null;
         initializeIterator(this.testField, this.cases, this.defaultCase);
     }
 
     @Override
     public Item next() {
         if (this.hasNext) {
-            this.matchingIterator.open(this.currentDynamicContextForLocalExecution);
             Item nextItem = this.matchingIterator.next();
             this.matchingIterator.close();
-            this.hasNext = false;
+            this.hasNext = this.matchingIterator.hasNext();
```

Comment:
This is a fix to make sure that it forwards *all* items, not just one.

Reply:
This line also looks quite interesting. Line 59 closes matching iterator and a closed iterator always has its hasNext set to false. So line 60 is equivalent to this.hasNext = false.

Furthermore, this makes the if check at line 61 redundant as we know !this.hasNext will always evaluate to true.

Reply:
This was a bug, the first close call should have been removed as well.

Reply:
Glad to help. This looks very clean now